File: src/lib/ai/computer-classify.ts
export type ComputerRisk = 'read' | 'write';

const READ_METHODS = new Set([
    'computer_screenshot',
    'computer.screenshot',
    'computer_screen_size',
    'computer.screen_size',
    'computer_cursor_position',
    'computer.cursor_position',
    'computer_find',
    'computer.find',
]);

export function classifyComputerAction(toolName: string): ComputerRisk {
    return READ_METHODS.has(toolName) ? 'read' : 'write';
}

const WRITE_DESCRIPTIONS: Record<string, (args: Record<string, unknown>) => string> = {
    computer_mouse_move: (a) => `Move mouse to (${a.x}, ${a.y})`,
    computer_left_click: (a) => a.x !== undefined ? `Left-click at (${a.x}, ${a.y})` : 'Left-click at current position',
    computer_right_click: (a) => a.x !== undefined ? `Right-click at (${a.x}, ${a.y})` : 'Right-click at current position',
    computer_middle_click: (a) => a.x !== undefined ? `Middle-click at (${a.x}, ${a.y})` : 'Middle-click at current position',
    computer_double_click: (a) => a.x !== undefined ? `Double-click at (${a.x}, ${a.y})` : 'Double-click at current position',
    computer_left_click_drag: (a) => `Drag from (${a.x1},${a.y1}) to (${a.x2},${a.y2})`,
    computer_type: (a) => {
        const text = String(a.text ?? '');
        const preview = text.length > 6 ? text.slice(0, 6) + '...' : text;
        return `Type "${preview}" (${text.length} chars)`;
    },
    computer_key: (a) => `Press key: ${a.key}`,
    computer_scroll: (a) => `Scroll ${a.direction} (${a.clicks ?? 3} clicks)`,
};

export function describeComputerAction(toolName: string, args: Record<string, unknown>): string {
    const fn = WRITE_DESCRIPTIONS[toolName];
    if (fn) return fn(args);
    return `${toolName}(${JSON.stringify(args)})`;
}

File: src-tauri/src/computer_classify.rs
use serde_json::Value;

#[derive(Debug, Clone, PartialEq, Eq)]
pub enum ComputerRisk {
    Read,
    Write,
}

impl ComputerRisk {
    pub fn as_str(&self) -> &'static str {
        match self {
            ComputerRisk::Read => "read",
            ComputerRisk::Write => "write",
        }
    }
}

pub fn classify(method: &str, _params: &Value) -> ComputerRisk {
    match method {
        "computer.screenshot"
        | "computer_screenshot"
        | "computer.screen_size"
        | "computer_screen_size"
        | "computer.cursor_position"
        | "computer_cursor_position"
        | "computer.find"
        | "computer_find" => ComputerRisk::Read,
        _ => ComputerRisk::Write,
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn read_methods_are_read() {
        for m in [
            "computer_screenshot",
            "computer_screen_size",
            "computer_cursor_position",
            "computer_find",
        ] {
            assert_eq!(classify(m, &json!({})), ComputerRisk::Read, "{}", m);
        }
    }

    #[test]
    fn write_methods_are_write() {
        for m in [
            "computer_mouse_move",
            "computer_left_click",
            "computer_right_click",
            "computer_double_click",
            "computer_middle_click",
            "computer_left_click_drag",
            "computer_type",
            "computer_key",
            "computer_scroll",
        ] {
            assert_eq!(classify(m, &json!({})), ComputerRisk::Write, "{}", m);
        }
    }

    #[test]
    fn unknown_method_defaults_write() {
        assert_eq!(classify("computer_format_drive", &json!({})), ComputerRisk::Write);
    }
}
